src/network/server.rs

```diff
@@ -1008,8 +1009,67 @@ impl HubService for MyHubService {
                 events
             })
             .collect();
-        let combined_events: Vec<HubEvent> =
+        let mut combined_events: Vec<HubEvent> =
```

Comment:
Can you combine all the filtering into one pass? It's more efficient and I don't think it'll be messier.